### PR TITLE
Add a transformer to use Gelu approximation for cuda provider

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: Add a level 2 transformer to convert Gelu or AddGeluFusion to FastGelu op.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Gelu cuda implementation is slow. Using approximation, the inference performance could significantly be improved.

### CODE REVIEW

**Comment by pranavsharma** [comment]
1. Curious if we need a separate op FastGelu for the approximation. Would it make sense to modify the Gelu op to take an attribute called 'mode' denoting the mode in which it operates (approximate vs exact)? Then we can potentially control this through a session option that says whether to run in approx mode or not. In this case we don't even need a separate transformer.
2. If the above doesn't make sense, do you think this qualifies to be a level 4 transformer? Why not level 2?

**Comment by tianleiwu** [comment]
> 1. Curious if we need a separate op FastGelu for the approximation. Would it make sense to modify the Gelu op to take an attribute called 'mode' denoting the mode in which it operates (approximate vs exact)? Then we can potentially control this through a session option that says whether to run in approx mode or not. In this case we don't even need a separate transformer.

It is nice to merge FastGelu with Gelu using an attribute. That could be a work item in the future. 

Using transformer has the following benefits:
(1) User could export optimized graph with approximation.
(2) No need a new session option. Currently, this is the only op that has approximation, it is not worth to introduce a new option right now.
(3) Extra check on input data type and shape. Currently, Gelu and FastGelu does not support same set of data types and input shape constraint is difference. So we need transformer to do some validation before conversion.

> 2. If the above doesn't make sense, do you think this qualifies to be a level 4 transformer? Why not level 2?

We add a new level because user could choose whether to use approximation or not. Note that level 2 has fusion of sub-graph to Gelu op. Level 2 is non-approximation optimization.

The concept is approximation will change output, thus it could impact the accuracy of result. So it is last resort in optimization. That is the reason I put it to a level higher than other optimizations.

File: onnxruntime/core/optimizer/gelu_approximation.cc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  for (const auto& input_arg : node.InputDefs()) {
+    if (std::find(supported_data_types.begin(), supported_data_types.end(),
+                  *(input_arg->Type())) == supported_data_types.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static bool CheckInputShape(Graph& graph, Node& node, const NodeArg& input, const NodeArg& bias) {
```
Comment (pranavsharma) [inline onnxruntime/core/optimizer/gelu_approximation.cc]: 1. You don't seem to be using ```graph``` in the function.
2. Can Node be passed using const ref?

File: onnxruntime/core/session/abi_session_options.cc [hunk 2]
```diff
@@ -130,7 +130,7 @@ ORT_API_STATUS_IMPL(OrtApis::SetSessionGraphOptimizationLevel, _In_ OrtSessionOp
       options->value.graph_optimization_level = onnxruntime::TransformerLevel::Level2;
       break;
     case ORT_ENABLE_ALL:
-      options->value.graph_optimization_level = onnxruntime::TransformerLevel::Level3;
```
Comment (pranavsharma) [inline onnxruntime/core/session/abi_session_options.cc]: How do users enable level 3 after this change if they don't want level 4?

File: onnxruntime/core/session/inference_session.cc [hunk 3]
```diff
@@ -1149,6 +1149,10 @@ void InferenceSession::AddPredefinedTransformers(GraphTransformerManager& transf
   if ((graph_optimization_level >= TransformerLevel::Level3) || !custom_list.empty()) {
     add_transformers(TransformerLevel::Level3);
   }
+
+  if ((graph_optimization_level >= TransformerLevel::Level4) || !custom_list.empty()) {
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.cc]: Should TransformerLevel::Level4 be replaced with MaxLevel?

**Comment by askhade** [comment]
> > 1. Curious if we need a separate op FastGelu for the approximation. Would it make sense to modify the Gelu op to take an attribute called 'mode' denoting the mode in which it operates (approximate vs exact)? Then we can potentially control this through a session option that says whether to run in approx mode or not. In this case we don't even need a separate transformer.
> 
> It is nice to merge FastGelu with Gelu using an attribute. That could be a work item in the future.
> 
> Using transformer has the following benefits:
> (1) User could export optimized graph with approximation.
> (2) No need a new session option. Currently, this is the only op that has approximation, it is not worth to introduce a new option right now.
> (3) Extra check on input data type and shape. Currently, Gelu and FastGelu does not support same set of data types and input shape constraint is difference for AddGeluFusion and FastGelu. So we need transformer to do some validation before conversion.
> 
> > 1. If the above doesn't make sense, do you think this qualifies to be a level 4 transformer? Why not level 2?
> 
> We add a new level because user could choose whether to use approximation or not. Note that level 2 has fusion of sub-graph to Gelu op. Level 2 is non-approximation optimization.
> 
> The concept is approximation will change output, thus it could impact the accuracy of result. So it is last resort in optimization. That is the reason I put it to a level higher than other optimizations.

We should avoid adding new levels for every new case. Since you mentioned right now this is the only op which qualifies for approximation - then is it worth it to introduce a new level for this (and add complexity) ?
My suggestion is - we should either add an attribute "mode" to the existing op or add this transformer to existing level2.

File: onnxruntime/core/session/inference_session.cc (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/core/session/inference_session.cc]: No. Considering the case that someone add level 5 in the future.

File: onnxruntime/core/session/abi_session_options.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/core/session/abi_session_options.cc]: It is not able to do that unless we change public API, current design has only three buckets. Internally, we could have many levels that mapped to these 3 buckets.

**Comment by tianleiwu** [comment]
> My suggestion is - we should either add an attribute "mode" to the existing op or add this transformer to existing level2.
I could add this transformer to existing level 3. 
Not level 2 because level 2 has Gelu fusion. Put it to level 3 so that user could choose not using approximation and only Gelu fusion.

**Comment by pranavsharma** [comment]
> > My suggestion is - we should either add an attribute "mode" to the existing op or add this transformer to existing level2.
> 
> I could add this transformer to existing level 2 after discussion with team members.

FYI - we were discussing this in the scrum today and looks like there are 2 other ops that could benefit from a session option to turn on/off approximate vs exact behavior.
With regards to level 2, if I turn on level 2, I'll get the approximation even if I don't want to, right?

**Comment by yufenglee** [comment]
That's good, so we should take the option to add a new opt level and adjust the API.

---
In reply to: [558876805](https://github.com/microsoft/onnxruntime/pull/2480#issuecomment-558876805) [](ancestors = 558876805)

File: include/onnxruntime/core/optimizer/graph_transformer_level.h [hunk 4]
```diff
@@ -12,9 +12,8 @@ enum class TransformerLevel : int {
   Level1,
   Level2,
   Level3,
-  // Convenience enum to always get the max available value.
-  // This way when we add more levels code which iterates over this enum does not need to change.
-  MaxTransformerLevel
+  // The max level should always be same as the last level.
+  MaxLevel = Level3
```
Comment (yufenglee) [inline include/onnxruntime/core/optimizer/graph_transformer_level.h]: We should add a new level here.

File: onnxruntime/core/optimizer/graph_transformer_utils.cc [hunk 5]
```diff
@@ -130,6 +131,7 @@ std::vector<std::unique_ptr<GraphTransformer>> GenerateTransformers(TransformerL
 
       std::unordered_set<std::string> cuda_execution_providers = {onnxruntime::kCudaExecutionProvider};
       transformers.emplace_back(onnxruntime::make_unique<AddGeluFusion>(cuda_execution_providers));
+      transformers.emplace_back(onnxruntime::make_unique<GeluApproximation>(cuda_execution_providers));
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/graph_transformer_utils.cc]: I think we should either add a new opt level or replace the AddGeluFusion GPU kernel with GeluApproximation kernel. 
If you take the latter option, you don't need to register a transformer here. You just need to replace the GPU kernel. I have a PR [https://github.com/microsoft/onnxruntime/pull/2487](https://github.com/microsoft/onnxruntime/pull/2487) that change AddGeluFusion to have same schema as the FastGelu.

Comment (tianleiwu) [reply onnxruntime/core/optimizer/graph_transformer_utils.cc]: After more discussion, here is a summary:
(1) It is nice to have a session option to enable or disable approximation. That option could enable or disable this transformer. However, we think it is not worth the effort to add it right now.
(2) Another option is we add this flag to a runtime configuration (to be built) that could saved with model. This feature is not available right now.

I will add this to level 2 transformer for now. If user does need non-approximated version, he/she could use our python script to generate an optimized BERT model, and disable online optimization in inference.

In the future, we could evaluate user requirements to see whether it need to add an option to enable or disable approximation.

**Comment by tianleiwu** [comment]
> FYI - we were discussing this in the scrum today and looks like there are 2 other ops that could benefit from a session option to turn on/off approximate vs exact behavior.
> With regards to level 2, if I turn on level 2, I'll get the approximation even if I don't want to, right?

Let's add this to back log: create a session option to turn on/off approximation.

User still have way to uses non-approximated version for Bert model: use level 1 optimization, and optimize graph using our python script.

File: onnxruntime/core/optimizer/gelu_approximation.cc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  return false;
+}
+
+static bool CheckGeluInputShape(const NodeArg& input) {
+  const TensorShapeProto* input_shape = input.Shape();
+  return nullptr != input_shape && input_shape->dim_size() >= 1;
+}
+
+static bool IsCandidateNode(Node& node, const std::unordered_set<std::string>& compatible_providers) {
```
Comment (pranavsharma) [inline onnxruntime/core/optimizer/gelu_approximation.cc]: Can Node by passed by const-ref here?

Comment (tianleiwu) [reply onnxruntime/core/optimizer/gelu_approximation.cc]: Will change it later.

**Pull request merged**